zapcore/write_syncer.go

```diff
@@ -75,6 +78,96 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	sync.Mutex
+	bufferWriter *bufio.Writer
+	ticker       *time.Ticker
+}
+
+// defaultBufferSize sizes the buffer associated with each WriterSync.
+const defaultBufferSize = 256 * 1024
+
+// defaultFlushInterval means the default flush interval
+const defaultFlushInterval = 30 * time.Second
+
+// CloseFunc should be called when the caller exits to clean up buffers.
+type CloseFunc func() error
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// if bufferSize = 0, we set it to defaultBufferSize
+// if flushInterval = 0, we set it to defaultFlushInterval
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (WriteSyncer, CloseFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	closefunc := func() error {
+		cancel()
+		return ws.Sync()
+	}
+
+	if _, ok := ws.(*bufferWriterSyncer); ok {
+		// no need to layer on another buffer
+		return ws, closefunc
+	}
+
+	if bufferSize == 0 {
+		bufferSize = defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = defaultFlushInterval
+	}
+
+	ticker := time.NewTicker(flushInterval)
+
+	ws = &bufferWriterSyncer{
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+		ticker:       ticker,
+	}
+
+	// flush buffer every interval
+	// we do not need to exit this goroutine until closefunc called explicitly
+	go func() {
+		select {
+		case <-ticker.C:
+			// the background goroutine just keep syncing
+			// until the close func is called.
+			_ = ws.Sync()
```

Comment:
Sorry if my comment is out of place here;

Is there a good way to collect these errors? If WriteSyncer makes a network call, like in a service I am currently writing, then these errors get lost which make it hard to debug

Reply:
The timer loop here is just another way to sync data. We mostly call Write() to sync data directly, which would collect these errors.

Reply:
I just noticed... this go routine will flush once after `flushInterval` then exit. There is no loop here.

Reply:
thanks...my mistake...fixed